### DISCUSSION TITLE
Source directory hiding build directory in analysis and autocomplete

### DISCUSSION THREAD

**Discussion by tristan957** [microsoft/pylance-release#4001] [Enhancement]
We have a Python bindings project for a C library. The Python project is https://github.com/hse-project/hse-python. We use Meson to build our Python bindings using its Cython support. Our build directory looks like:

```
builddir/hse3
├── hse.c
├── hse.cpython-310-x86_64-linux-gnu.so
├── hse.cpython-310-x86_64-linux-gnu.so.p
│   └── meson-generated_.._hse.c.o
├── hse.pxd
├── hse.pyi
├── hse.pyi.pp
├── hse.pyx
├── __init__.py
├── limits.c
├── limits.cpython-310-x86_64-linux-gnu.so
├── limits.cpython-310-x86_64-linux-gnu.so.p
│   └── meson-generated_.._limits.c.o
├── limits.pxd
├── limits.pyi
├── limits.pyi.pp
├── limits.pyx
├── __pycache__
│   └── __init__.cpython-310.pyc
├── py.typed
├── version.c
├── version.cpython-310-x86_64-linux-gnu.so
├── version.cpython-310-x86_64-linux-gnu.so.p
│   └── meson-generated_.._version.c.o
├── version.pxd
├── version.pyi
├── version.pyi.pp
└── version.pyx
```

And the source dir looks like:

```
$sourcedir/hse3
hse3
├── hse.in.pyi
├── hse.in.pyx
├── hse.pxd
├── __init__.py
├── limits.in.pyi
├── limits.in.pyx
├── limits.pxd
├── meson.build
├── py.typed
├── version.in.pyi
├── version.in.pyx
└── version.pxd
```

I believe because the source directory is called `hse3` and has an `__init__.py`, it keeps Pylance from understanding that the `builddir/hse3` is the actual directory to resolve.

![image](https://user-images.githubusercontent.com/11936772/193704565-520f79db-3ea4-43c4-8423-19c8165be254.png)

```
"hse" is unknown import symbol
```

I have tried playing with the settings in my workspace like so:

```jsonc
{
    "python.analysis.extraPaths": [
      "path/to/builddir",
    ],
    "python.analysis.exclude": [
      "path/to/sourcedir",
    ],
    "python.analysis.include": [
      "path/to/builddir",
    ],
    "python.autoComplete.extraPaths": [
      "path/to/builddir"
    ],
}
```

But nothing is working, so essentially my feature request is to provide a way to mask a directory if at all possible.

**Comment by rchiodo** [comment]
I can repro the issue.

Personally, I think the 'exclude' setting isn't named very well. It only excludes things as being user files. It doesn't exclude them from being imported. 

However, we might be able to adjust the import resolver to favor non excluded files first.

**Comment by tristan957** [comment]
> However, we might be able to adjust the import resolver to favor non excluded files first.

This would work for my case.

**Comment by rchiodo** [comment]
Next week's release should have the 'fix' for this issue in it. In your example, you'd do this to change the import sort order:

```jsonc
{
    "python.analysis.exclude": [
        "sourcedir/**"
    ],
    "python.analysis.extraPaths": [
        "builddir"
    ]
}
```

**Comment by tristan957** [comment]
@rchiodo thanks so much for solving this. Can't wait for next week!

**Comment by bschnurr** [comment]
This issue has been fixed in prerelease version 2023.2.41, which we've just released. You can find the changelog here: [CHANGELOG.md](https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#2023241-22-February-2023-PreRelease)

**Comment by rchiodo** [comment]
Turns out our 'fix' broke something else. This would need something custom then. Turning into a discussion
